Sources/MongoSwift/APM.swift

```diff
@@ -1,14 +1,109 @@
+    func handleServerHeartbeatEvent(_ event: ServerHeartbeatEvent)
+}
+
+public extension SDAMEventHandler {
+    /// Handle a `TopologyEvent`.
+    /// Implementing this method is optional. If no implementation is provided, `TopologyEvent`s will be ignored.
+    func handleTopologyEvent(_: TopologyEvent) {}
+
+    /// Handle a `ServerHeartbeatEvent`.
+    /// Implementing this method is optional. If no implementation is provided, `ServerHeartbeatEvent`s will be ignored.
+    func handleServerHeartbeatEvent(_: ServerHeartbeatEvent) {}
+}
+
+/// A protocol for events that are directly consumable by users to implement.
+private protocol Publishable {
+    func publish(to client: MongoClient)
+}
+
+/// A protocol for monitoring events to implement, indicating that they can be initialized from a libmongoc event
+/// and that they can be packaged into a type which can be published.
+private protocol MongoSwiftEvent {
+    associatedtype MongocEventType: MongocEvent
+    associatedtype PublishableEventType: Publishable
+
+    init(mongocEvent: MongocEventType)
+
+    func toPublishable() -> PublishableEventType
+}
+
+/// A protocol for libmongoc event wrappers to implement.
+private protocol MongocEvent {
+    init(_ eventPtr: OpaquePointer)
+
+    var context: UnsafeMutableRawPointer? { get }
+}
+
+/// A command monitoring event.
+public enum CommandEvent: Publishable {
+    /// An event published when a command starts.
+    case started(CommandStartedEvent)
+
+    /// An event published when a command succeeds.
+    case succeeded(CommandSucceededEvent)
+
+    /// An event published when a command fails.
+    case failed(CommandFailedEvent)
+
+    private var event: CommandEventProtocol {
+        switch self {
+        case let .started(event):
+            return event
+        case let .succeeded(event):
+            return event
+        case let .failed(event):
+            return event
+        }
+    }
+
+    fileprivate func publish(to client: MongoClient) {
+        guard let handler = client.commandEventHandler else {
+            return
+        }
+        handler.handleCommandEvent(self)
+    }
+
+    /// The name of the command that generated this event.
+    public var commandName: String {
+        return self.event.commandName
+    }
+
+    /// The driver generated request id.
+    public var requestId: Int64 {
+        return self.event.requestId
+    }
+
+    /// The driver generated operation id. This is used to link events together such
+    /// as bulk write operations.
+    public var operationId: Int64 {
+        return self.event.operationId
+    }
+
+    /// The address of the server the command was run against.
+    public var serverAddress: Address {
+        return self.event.serverAddress
+    }
 }
 
 /// A protocol for command monitoring events to implement, specifying the command name and other shared fields.
-public protocol MongoCommandEvent: MongoEvent {
+private protocol CommandEventProtocol {
```

Comment:
This protocol was retained privately to reduce the boilerplate in `CommandEvent`.